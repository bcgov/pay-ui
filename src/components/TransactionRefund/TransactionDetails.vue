<template>
  <v-card flat>
    <v-card-title class="card-title">
      <v-icon
        class="pr-3"
        color="link"
      >
        mdi-file-document
      </v-icon>
      Transaction Details
    </v-card-title>
    <v-card-text
      class="mt-5"
    >
      <v-row no-gutters>
        <v-col class="col-12 col-sm-12 ">
          <v-row>
            <v-col class="col-6 col-sm-3 font-weight-bold">
              Transaction Date
            </v-col>
            <v-col class="col-6 col-sm-9">
              {{formatUtcToPacificDate(transactionData.invoiceCreatedOn, 'MMMM DD, YYYY h:mm A')}}
            </v-col>
          </v-row>
          <v-row>
            <v-col class="col-6 col-sm-3 font-weight-bold">
              Receipt
            </v-col>
            <v-col class="col-6 col-sm-9">
              <div
                v-if="transactionData.invoiceStatusCode === InvoiceStatus.COMPLETED || transactionData.invoiceStatusCode === InvoiceStatus.PAID"
                class="receipt"
                @click="downloadReceipt()"
              >
                <v-icon
                  color="primary"
                >
                  mdi-file-pdf-outline
                </v-icon>
                <span>
                  Receipt
                </span>
              </div>
            </v-col>
          </v-row>
          <v-row>
            <v-col class="col-12">
              <v-simple-table>
                <template v-slot:default>
                  <thead>
                    <tr>
                      <th class="text-left">
                        Application
                      </th>
                      <th class="text-left">
                        Type
                      </th>
                      <th class="text-left">
                        Number
                      </th>
                      <th class="text-left">
                        Detail
                      </th>
                    </tr>
                  </thead>
                  <tbody>
                    <tr>
                      <td>
                        {{transactionData.applicationName}}
                      </td>
                      <td>
                        {{transactionData.applicationType}}
                      </td>
                      <td>
                        {{transactionData.businessIdentifier}}
                      </td>
                      <td>
                        <template v-if="transactionData.applicationDetails">
                          <div v-for="(detail, index) in transactionData.applicationDetails" :key="index">
                            {{ detail.label }} {{ detail.value }}
                          </div>
                        </template>
                      </td>
                    </tr>
                  </tbody>
                </template>
              </v-simple-table>
            </v-col>
          </v-row>
          <v-row>
            <v-col class="col-6 col-sm-3 font-weight-bold py-0">
              Identifier
            </v-col>
          </v-row>
          <v-row>
            <v-col class="col-12">
              <v-simple-table>
                <template v-slot:default>
                  <thead>
                    <tr>
                      <th class="text-left">
                        Transaction ID
                      </th>
                      <th class="text-left">
                        Invoice Reference ID
                      </th>
                    </tr>
                  </thead>
                  <tbody>
                    <tr>
                      <td>
                        {{transactionData.invoiceId}}
                      </td>
                      <td>
                        {{transactionData.invoiceReferenceId}}
                      </td>
                    </tr>
                  </tbody>
                </template>
              </v-simple-table>
            </v-col>
          </v-row>
        </v-col>
      </v-row>
    </v-card-text>
  </v-card>
</template>

<script lang="ts">
import { defineComponent, PropType } from '@vue/composition-api'
import CommonUtils from '@/util/common-util'
import { TransactionData } from '@/models/transaction-refund'
import PaymentService from '@/services/payment.services'
import ConfigHelper from '@/util/config-helper'
import { InvoiceStatus, SessionStorageKeys } from '@/util/constants'

export default defineComponent({
  name: 'TransactionDetails',
  props: {
    transactionData: {
      type: Object as PropType<TransactionData>,
      required: true
    }
  },
  setup (props) {
    async function downloadReceipt () {
      const currentAccount = JSON.parse(ConfigHelper.getFromSession(SessionStorageKeys.CurrentAccount || '{}'))
      const receipt = await PaymentService.postReceipt(
        props.transactionData.invoiceId,
        currentAccount.id,
        CommonUtils.formatDisplayDate(props.transactionData.invoiceCreatedOn)
      )
      const filename = `bcregistry-receipts-${props.transactionData.invoiceId}.pdf`
      CommonUtils.fileDownload(receipt.data, filename, 'application/pdf')
    }

    return {
      formatDisplayDate: CommonUtils.formatDisplayDate,
      formatUtcToPacificDate: CommonUtils.formatUtcToPacificDate,
      downloadReceipt,
      InvoiceStatus
    }
  }
})
</script>

<style lang="scss" scoped>
@import '$assets/scss/theme.scss';
@import '$assets/scss/actions.scss';

.card-title {
  background-color: $app-lt-blue;
  justify-content: left;
  height: 75px;
  font-weight: bold;
  font-size: 1.125rem;
}

.v-data-table > .v-data-table__wrapper > table > thead > tr > th {
  border-right: 1px solid #E0E0E0;
}

.v-data-table > .v-data-table__wrapper > table > tbody > tr:hover {
  background-color: transparent !important;
}

.v-data-table > .v-data-table__wrapper > table > thead > tr > th:first-child,
.v-data-table > .v-data-table__wrapper > table > tbody > tr > td:first-child {
  padding-left: 0 !important;
}

.v-data-table > .v-data-table__wrapper > table > thead > tr > th:not(:first-child),
.v-data-table > .v-data-table__wrapper > table > tbody > tr > td:not(:first-child) {
  padding-left: 16px !important;
}

.v-data-table > .v-data-table__wrapper > table > tbody > tr > td {
  padding-top: 16px;
  vertical-align: top;
}
.v-application ol, .v-application ul {
  padding-left: 16px;
}
.receipt {
  cursor: pointer;
  color: var(--v-primary-base);
}

</style>
