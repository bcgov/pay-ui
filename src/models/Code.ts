// To handle codes tables such as SuspensionReasonCode, etc
export interface Code {
    code?: string;
    description?: string;
}

export interface PaymentMethod {
    code: string;
    description: string;
    partialRefund?: boolean;
    refundableStatuses?: string;
}
