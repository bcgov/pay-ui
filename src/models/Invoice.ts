// Model for storing invoice of a routing slip
export interface LineItem {
  description?: string
  filingFees?: number
  futureEffectiveFees?: number
  gst?: number
  id?: number
  priorityFees?: number
  pst?: number
  quantity?: number
  serviceFees?: number
  statusCode?: string
  total?: number
  waivedBy?: string
  waivedFees?: number
}

export interface Reference {
  id?: number
  invoiceNumber?: string
  statusCode?: string
}

export interface Detail {
  label?: string
  value?: string
}

export interface Invoice {
  businessIdentifier?: string
  corpTypeCode?: string
  createdBy?: string
  createdOn?: string
  createdName?: string
  id?: number
  lineItems?: LineItem[]
  detail?: Detail[]
  paid?: number,
  paymentAccount?: {
    accountId?: string
    accountName?: string
    billable?: boolean
  }
  folioNumber?: string
  paymentMethod?: string
  references?: Reference[]
  refund?: number
  routingSlip?: string
  serviceFees?: string
  statusCode?: string
  total?: number,
  details?: Detail[]
}

export interface InvoiceList {
  consInvNumber?: string
  invoiceNumber: string
  invoices: Invoice[],
  paymentMethod: string
  paymentSystem: string
  statusCode: string
  invoiceAmount: number
  paidAmount: number
}

export interface InvoiceListResponse {
  items: InvoiceList[]
  limit: number
  page: number
  total: number
}

// For displaying transaction data table
export interface InvoiceDisplay {
  createdOn?: string
  invoiceNumber?: string
  statusCode?: string
  total?: number
  createdName?: string
  createdBy?: string
  description?: string[]
  id?: number
}
