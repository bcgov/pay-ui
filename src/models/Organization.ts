import { Contact } from './contact'

export interface Organization {
  id?: number
}

export interface Organizations{
  orgs : Organization[]
}

export interface PADInfo {
  bankAccountNumber: string
  bankInstitutionNumber: string
  bankTransitNumber: string
  isTOSAccepted?: boolean
  isAcknowledged?: boolean
}

export interface CFSAccountDetails {
  bankAccountNumber: string
  bankInstitutionNumber: string
  bankTransitNumber: string
  cfsAccountNumber: string
  cfsPartyNumber: string
  cfsSiteNumber: string
  status: string
}

export interface GLInfo {
  client: string
  responsibilityCentre: string
  serviceLine: string
  stob: string
  projectCode: string
}

export interface OrgPaymentDetails {
  accountId: string
  accountName: string
  bcolAccount: string
  bcolUserId: string
  cfsAccount: CFSAccountDetails
  obCredit: string
  padCredit: string
  paymentMethod: string
  statementNotificationEnabled: true
  padTosAcceptedBy:string
  padTosAcceptedDate:string
  futurePaymentMethod: string
  revenueAccount?: GLInfo
  eftEnable?: boolean
}

export interface RoleInfo {
  icon: string
  name: string
  desc: string
  displayName: string
  displayOrder: number
  label: string
}

export interface AddUserBody {
  username: string
  password: string
  selectedRole?: RoleInfo
  membershipType: string
}

export interface AddUsersToOrgBody {
    users: AddUserBody[]
    orgId: string
}

export interface BulkUserResponse {
  httpStatus: number
  username: string
  created?: string
  error?: string
  firstname?: string
  type?: string
}

export interface BulkUserResponseBody {
  users: BulkUserResponse[]
}

export enum MembershipStatus {
  'Active' = 'ACTIVE',
  'Inactive' = 'INACTIVE',
  'Rejected' = 'REJECTED',
  'Pending' = 'PENDING_APPROVAL',
  'PendingStaffReview'='PENDING_STAFF_REVIEW'
}

export enum MembershipType {
  'Admin' = 'ADMIN',
  'Coordinator' = 'COORDINATOR',
  'User' = 'USER',
  'Staff' = 'STAFF',
  'ExternalStaff' = 'EXTERNAL_STAFF'
}

export interface UserTerms {
    isTermsOfUseAccepted: boolean
    termsOfUseAcceptedVersion: string
}

export interface User {
    firstname: string;
    lastname: string;
    username?: string;
    modified?: Date
    userTerms?: UserTerms
    contacts?: Contact[]
    email?: string,
    loginSource?: string
    id?: number
    keycloakGuid?: string
    verified?:boolean
}

export interface Member {
  id: number
  membershipTypeCode: MembershipType
  membershipStatus: MembershipStatus
  roleDisplayName?:string
  user: User
}
