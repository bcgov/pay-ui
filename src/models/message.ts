export interface Message {
  text: string
  type: 'success' | 'error'
  show: boolean
}
