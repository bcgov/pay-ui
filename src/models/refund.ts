export interface RefundRevenueType {
  paymentLineItemId: string
  refundAmount: number
  refundType: string
}

export interface RefundRequest {
  reason: string
  refundRevenue?: RefundRevenueType[]
}

export interface EftRefundBaseRequest {
  shortNameId: number
  refundAmount: number
  refundEmail: string
  comment?: string
}

export interface EftRefundRequest extends EftRefundBaseRequest{
  casSupplierNum: string
  casSupplierSite: string
}

export interface EftRefundChequeRequest extends EftRefundBaseRequest{
  entityName: string
  addressCity: string
  addressCountry: string
  addressRegion: string
  deliveryInstructions?: string
  postalCode: string
  streetAddress: string
  streetAddressAdditional?: string
}
