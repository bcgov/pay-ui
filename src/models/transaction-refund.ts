import { Detail } from './Invoice'

export interface PaymentData {
  accountName: string | null
  folioNumber: string | null
  initiatedBy: string | null
  paymentMethod: string | null
  paymentStatus: string | null
  totalTransactionAmount: number
}

export interface TransactionData {
  invoiceId: number | null
  transactionDate: string | null
  invoiceReferenceId: string | null
  invoiceStatusCode: string | null
  invoiceCreatedOn: string | null
  transactionAmount: number
  applicationName: string | null
  applicationType: string | null
  businessIdentifier: string | null
  applicationDetails: Detail[] | null
}

export enum RefundType {
  FULL_REFUND = 'Full Refund',
  PARTIAL_REFUND = 'Partial Refund'
}

export interface RefundLineItem {
  id: number | null
  description: string | null
  filingFees?: number | null
  serviceFees?: number | null
  priorityFees?: number | null
  total?: number | null
  futureEffectiveFees?: number | null
  filingFeesRequested?: number | string | null
  serviceFeesRequested?: number | string | null
  priorityFeesRequested?: number | string | null
  futureEffectiveFeesRequested?: number | string | null
  refundEntireItemRequested?: boolean | null
}

export interface RefundFormData {
  refundType: RefundType | null
  refundLineItems: RefundLineItem[] | null
  totalRefundAmount: number | null
  refundMethod: string | null
  notificationEmail: string | null
  reasonsForRefund: string | null
  staffComment: string | null
  requestedBy: string | null
  requestedTime: string | null
}

export enum RefundRequestStage {
  REQUEST_FORM = 'Request Form',
  DATA_VALIDATED = 'DATA Validated'
}
