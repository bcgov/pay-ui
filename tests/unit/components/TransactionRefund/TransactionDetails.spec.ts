import { createLocalVue, shallowMount } from '@vue/test-utils'
import Vuetify from 'vuetify'

import TransactionDetails from '@/components/TransactionRefund/TransactionDetails.vue'
import CommonUtils from '@/util/common-util'
import { TransactionData } from '@/models/transaction-refund'

describe('TransactionDetails.vue', () => {
  const localVue = createLocalVue()
  const vuetify = new Vuetify({})

  const sampleData: TransactionData = {
    invoiceId: 1234,
    transactionDate: '2025-01-02T00:00:00Z',
    invoiceStatusCode: 'PAID',
    invoiceCreatedOn: '2025-01-03T00:00:00Z',
    invoiceReferenceId: 'INV-123',
    transactionAmount: 100,
    applicationName: 'Name Request',
    applicationType: 'NR',
    businessIdentifier: 'NR 1234567',
    applicationDetails: [
      { label: 'NR Number:', value: 'NR_NUMBER' },
      { label: 'Name Choices:', value: '' }
    ]
  }

  beforeEach(() => {
    vi.resetModules()
    vi.clearAllMocks()
  })

  it('renders component', () => {
    vi.spyOn(CommonUtils, 'formatUtcToPacificDate').mockReturnValue('Jan 02, 2025')

    const wrapper = shallowMount(TransactionDetails, {
      localVue,
      vuetify,
      propsData: { transactionData: sampleData },
      directives: { can () { } }
    })

    expect(wrapper.findComponent(TransactionDetails).exists()).toBe(true)
    expect(wrapper.text()).toContain('Transaction Details')
    expect(wrapper.text()).toContain('mdi-file-document')
    expect(wrapper.text()).toContain('Name Request')
    expect(wrapper.text()).toContain('NR')
    expect(wrapper.text()).toContain('NR 1234567')
    expect(wrapper.text()).toContain('1234')
    expect(wrapper.text()).toContain('INV-123')
    expect(wrapper.text()).toContain('Jan 02, 2025')
  })

  it('matches snapshot', () => {
    vi.spyOn(CommonUtils, 'formatUtcToPacificDate').mockReturnValue('Jan 02, 2025')

    const wrapper = shallowMount(TransactionDetails, {
      localVue,
      vuetify,
      propsData: { transactionData: sampleData },
      directives: { can () { } }
    })
    expect(wrapper.html()).toMatchSnapshot()
  })
})
