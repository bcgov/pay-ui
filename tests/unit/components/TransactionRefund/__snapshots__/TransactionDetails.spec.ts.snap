// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TransactionDetails.vue > matches snapshot 1`] = `
"<v-card-stub data-v-2bb8060d="" loaderheight="4" tag="div" flat="true">
  <v-card-title-stub data-v-2bb8060d="" tag="div" class="card-title">
    <v-icon-stub data-v-2bb8060d="" color="link" class="pr-3"> mdi-file-document </v-icon-stub> Transaction Details
  </v-card-title-stub>
  <v-card-text-stub data-v-2bb8060d="" tag="div" class="mt-5">
    <v-row-stub data-v-2bb8060d="" tag="div" nogutters="true">
      <v-col-stub data-v-2bb8060d="" tag="div" class="col-12 col-sm-12 ">
        <v-row-stub data-v-2bb8060d="" tag="div">
          <v-col-stub data-v-2bb8060d="" tag="div" class="col-6 col-sm-3 font-weight-bold"> Transaction Date </v-col-stub>
          <v-col-stub data-v-2bb8060d="" tag="div" class="col-6 col-sm-9"> Jan 02, 2025 </v-col-stub>
        </v-row-stub>
        <v-row-stub data-v-2bb8060d="" tag="div">
          <v-col-stub data-v-2bb8060d="" tag="div" class="col-6 col-sm-3 font-weight-bold"> Receipt </v-col-stub>
          <v-col-stub data-v-2bb8060d="" tag="div" class="col-6 col-sm-9">
            <div data-v-2bb8060d="" class="receipt">
              <v-icon-stub data-v-2bb8060d="" color="primary"> mdi-file-pdf-outline </v-icon-stub><span data-v-2bb8060d=""> Receipt </span>
            </div>
          </v-col-stub>
        </v-row-stub>
        <v-row-stub data-v-2bb8060d="" tag="div">
          <v-col-stub data-v-2bb8060d="" tag="div" class="col-12">
            <v-simple-table-stub data-v-2bb8060d="">
              <thead data-v-2bb8060d="">
                <tr data-v-2bb8060d="">
                  <th data-v-2bb8060d="" class="text-left"> Application </th>
                  <th data-v-2bb8060d="" class="text-left"> Type </th>
                  <th data-v-2bb8060d="" class="text-left"> Number </th>
                  <th data-v-2bb8060d="" class="text-left"> Detail </th>
                </tr>
              </thead>
              <tbody data-v-2bb8060d="">
                <tr data-v-2bb8060d="">
                  <td data-v-2bb8060d=""> Name Request </td>
                  <td data-v-2bb8060d=""> NR </td>
                  <td data-v-2bb8060d=""> NR 1234567 </td>
                  <td data-v-2bb8060d="">
                    <div data-v-2bb8060d=""> NR Number: NR_NUMBER </div>
                    <div data-v-2bb8060d=""> Name Choices: </div>
                  </td>
                </tr>
              </tbody>
            </v-simple-table-stub>
          </v-col-stub>
        </v-row-stub>
        <v-row-stub data-v-2bb8060d="" tag="div">
          <v-col-stub data-v-2bb8060d="" tag="div" class="col-6 col-sm-3 font-weight-bold py-0"> Identifier </v-col-stub>
        </v-row-stub>
        <v-row-stub data-v-2bb8060d="" tag="div">
          <v-col-stub data-v-2bb8060d="" tag="div" class="col-12">
            <v-simple-table-stub data-v-2bb8060d="">
              <thead data-v-2bb8060d="">
                <tr data-v-2bb8060d="">
                  <th data-v-2bb8060d="" class="text-left"> Transaction ID </th>
                  <th data-v-2bb8060d="" class="text-left"> Invoice Reference ID </th>
                </tr>
              </thead>
              <tbody data-v-2bb8060d="">
                <tr data-v-2bb8060d="">
                  <td data-v-2bb8060d=""> 1234 </td>
                  <td data-v-2bb8060d=""> INV-123 </td>
                </tr>
              </tbody>
            </v-simple-table-stub>
          </v-col-stub>
        </v-row-stub>
      </v-col-stub>
    </v-row-stub>
  </v-card-text-stub>
</v-card-stub>"
`;
